tpu-provisioner/internal/controller/synced_start_controller.go

```diff
@@ -0,0 +1,223 @@
+				return ctrl.Result{}, fmt.Errorf("setting owner reference: %w", err)
+			}
+			if err := r.Create(ctx, &corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      configMapName,
+					Namespace: req.Namespace,
+				},
+			}); err != nil {
+				if apierrors.IsAlreadyExists(err) {
+					return ctrl.Result{Requeue: true}, nil
+				}
+				return ctrl.Result{}, fmt.Errorf("creating ConfigMap: %w", err)
+			}
+		}
+		return ctrl.Result{}, fmt.Errorf("getting ConfigMap: %w", err)
+	}
+
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+
+	// If the ConfigMap has already been updated, return early.
+	if cm.Data[allStartedConfigMapKey] == "true" {
+		return ctrl.Result{}, nil
+	}
+
+	// List all Pods in the current restart of the JobSet.
+	var pods corev1.PodList
+	if err := r.List(ctx, &pods, client.MatchingLabels{
+		jobsetNameLabel:    js.Name,
+		jobsetRestartLabel: strconv.Itoa(int(js.Status.Restarts)),
+	}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("listing pods: %w", err)
+	}
+
+	threshold, ok := startedThreshold(js)
+	if !ok {
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: See if it is possible to add an index to .status.startTime
+	// to avoid the for loop.
+	var startedCount int32
+	var thresholdMet bool
+	for _, p := range pods.Items {
+		if p.Status.StartTime != nil {
+			startedCount++
+			if startedCount >= threshold {
+				thresholdMet = true
+				break
+			}
+		}
+	}
+
+	if !thresholdMet {
+		return ctrl.Result{}, nil
+	}
+
+	// Update the ConfigMap to indicate that all Pods have started.
+	cm.Data[allStartedConfigMapKey] = "true"
+
+	if err := ctrl.SetControllerReference(js, cm, r.Scheme); err != nil {
+		return ctrl.Result{}, fmt.Errorf("setting owner reference: %w", err)
+	}
+	if err := r.Update(ctx, cm); err != nil {
+		return ctrl.Result{}, fmt.Errorf("updating ConfigMap: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SyncedStartReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&jobset.JobSet{}).
+		Watches(&corev1.Pod{}, handler.EnqueueRequestsFromMapFunc(handler.MapFunc(jobsetForPod))).
+		Complete(r)
+	// TODO: Configure specific rate-limit settings if the default settings are not appropriate.
+}
+
+func jobsetForPod(_ context.Context, obj client.Object) []reconcile.Request {
+	labels := obj.GetLabels()
+	if labels == nil {
+		return nil
+	}
+
+	jobsetName, ok := labels[jobsetNameLabel]
+	if !ok {
+		return nil
+	}
+
+	return []reconcile.Request{
+		{
+			NamespacedName: types.NamespacedName{
+				Name:      jobsetName,
+				Namespace: obj.GetNamespace(),
+			},
+		},
+	}
+}
+
+func startedThreshold(js *jobset.JobSet) (int32, bool) {
+	cmName := js.Annotations[jobsetSyncedStartAnnotation]
+
+	var (
+		n     int32
+		found bool
+	)
+
+	for _, rj := range js.Spec.ReplicatedJobs {
+		for _, v := range rj.Template.Spec.Template.Spec.Volumes {
```

Comment:
We already exit the main reconcile loop early if the configMap name annotation doesn't exist on the JobSet, do we need to do this check here?

Unless there's another reason for it, I think we can just do:
```
podsExpected = 0
for each replictedJob:
  podsExpected += replicatedJob.replicas * replicatedJob.template.Paralellism
```

Reply:
I was thinking that there might be replicated jobs that are not a part of the synced-start set of pods.

Reply:
I see, sounds good